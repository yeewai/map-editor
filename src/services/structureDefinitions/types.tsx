export interface StructureDefinition {
    id: string,
    name: string,
    description?: string,
    width: number,
    length: number,
    kind: string,
    imageUrl: string,
    imageWidth?: number,
    imageHeight?: number,
    image?: any
}

export const emptyStructureDefintion = {
    id: "",
    name: "",
    width: 1,
    length: 1,
    kind: "",
    imageUrl: ""
}

export interface State {
    items: StructureDefinition[],
    itemsRequested: number,
    isFetching: boolean,
    hasFetched: boolean,
    error?: string
}
