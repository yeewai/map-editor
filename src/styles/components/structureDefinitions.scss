.card-li {
    // cursor: pointer;
    // // background: var(--svp-content-bg);
    // border: none;
    // color: black;//var(--svp-body-text);
    // max-width: 180px;
    // height: 180px;
    // max-height: 180px;
    margin: 10px;
    // text-align: center;
    display: inline-block;

    .card {
        width: 240px;
        height: 360px;
        box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
        transition: box-shadow 0.4s cubic-bezier(.25,.8,.25,1);
        border: none;

        &:hover, &:focus {
            box-shadow: 0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23);
            color: var(--svp-body-text);
            text-decoration: none;
            outline: none;
        }

        img {
            height: 160px;
        }

        .card-title, .card-subtitle {
            font-weight: 300;
        }

        .card-subtitle {
            font-size: .8rem;
        }
    }

}
