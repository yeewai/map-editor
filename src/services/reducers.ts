import { routerReducer } from 'react-router-redux';
import { combineReducers } from 'redux';
import { modalReducer } from 'ever-modal';
import { reducer as formReducer } from 'redux-form'

import { StateTree } from './types';
import structureDefinitions from './structureDefinitions';
import worlds from './worlds';

const rootReducer = combineReducers<StateTree>({
    router: routerReducer,
    modals: modalReducer,
    form: formReducer,
    structureDefinitions,
    worlds

});

export default rootReducer;
