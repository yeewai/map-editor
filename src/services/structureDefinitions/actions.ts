import { ActionCreator, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { StructureDefinition } from './types';
import { StateTree, Action } from 'services/types';
import * as api from 'api';

import { modalActions } from 'ever-modal';

export const requestStructureDefinitions: ActionCreator<Action> = () => ({
    type: 'structureDefinition/REQUEST'
});

export const receiveStructureDefinitions: ActionCreator<Action> = ( definitions: StructureDefinition[] )=> ({
    type: 'structureDefinition/RECEIVE',
    payload: definitions
});

export const setError: ActionCreator<Action> = (jqXHR: any, textStatus: string, errorThrown: any) => ({
    type: 'structureDefinition/SET_ERROR',
    payload: { jqXHR, textStatus, errorThrown }
});

//I have no idea what the last return type is so I set it to any.........
export const fetchStructureDefinitions: ActionCreator<ThunkAction<any, StateTree, void>> = () => {
    return (dispatch: Dispatch<StateTree>): any => {
        dispatch(requestStructureDefinitions());
        return api.getStructureDefinitions().then (
            resp => dispatch(receiveStructureDefinitions(resp)),
            (jqXHR, textStatus, errorThrown) =>{
                 dispatch(setError(jqXHR, textStatus, errorThrown))
             }
        );
    };
};

export const putStructureDefinition: ActionCreator<Action> = () => ({
    type: 'structureDefinition/PUT'
});

export const updateStructureDefinition: ActionCreator<ThunkAction<any, StateTree, void>> = (values) => {
    return (dispatch: Dispatch<StateTree>, getState: () => StateTree ) => {

        dispatch(putStructureDefinition());
        return api.updateStructureDefinition(values.id, values).then(
            resp => {
                dispatch(modalActions.hideModal());
                dispatch(fetchStructureDefinitions());
            },
            (jqXHR, textStatus, errorThrown) => dispatch(setError(jqXHR, textStatus, errorThrown))
        );
    };
}
