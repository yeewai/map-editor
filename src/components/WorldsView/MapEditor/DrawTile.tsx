import * as React from 'react';
import { connect } from "react-redux";
import { ListGroup, ListGroupItem, Alert } from 'reactstrap';
import moment from 'moment';
import _ from 'lodash';

import { TILE_SIZE } from 'conf';
import { OpenModalButton } from '@evercourse/ever-modal';

import { StateTree } from 'services/types';
import { worldTypes } from 'services/worlds';
import { structureDefinitionTypes } from 'services/structureDefinitions';

interface OwnProps {
    structureDefinition: structureDefinitionTypes.StructureDefinition,
    x: number,
    y: number
}

export type PropTypes = OwnProps;

const tileWidthHalf = TILE_SIZE/2;
const tileHeightHalf = (tileWidthHalf * Math.tan( 30 * Math.PI/180)); //Thanks geometry @.@

const getCoords = (x: number, y:number): {x: number, y: number} => {

    return {
        x: (x - y) * tileWidthHalf,
        y: (x + y) * tileHeightHalf
    };
}

const getCoordsString =(x: number, y:number): string => {
    const coords = getCoords(x, y);
    return `${coords.x},${coords.y}`
}

export class DrawTile extends React.Component<PropTypes> {
    shouldComponentUpdate(nextProps) {
        return this.props.structureDefinition !== nextProps.structureDefinition
            || this.props.x !== nextProps.x
            || this.props.y !== nextProps.y;
    }

    render() {
        const { structureDefinition, x, y } = this.props;
        if (!structureDefinition) { return (
            <polygon
                points={[getCoordsString(x, y),getCoordsString(x+1, y),getCoordsString(x+1, y+1),getCoordsString(x, y+1)].join(" ")}
                className="nullTile"  />
        ) }

        const width = getCoords(x+structureDefinition.width, y).x - getCoords(x,y+structureDefinition.length).x;
        const coords = getCoords(
            x,
            y
        );

        const ratio = width/structureDefinition.imageWidth;
        const height = structureDefinition.imageHeight * ratio;

        return (
            <use className="tile"
                width={ width }
                height={height}
                x={ coords.x - (width/2) }
                y={ coords.y - (height) + (tileHeightHalf*2)}
                xlinkHref={`#${structureDefinition.id}`}
            />
        )
    };
}

export default DrawTile;
