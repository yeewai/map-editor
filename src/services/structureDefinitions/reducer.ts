import { Action } from 'services/types';
import * as Types from './types';

export const defaultState: Types.State = {
    isFetching: false,
    hasFetched: false,
    error: undefined,
    items: [],
    itemsRequested: 0
}

const structureDefinitionReducer = (state = defaultState, action: Action ): Types.State => {
    switch(action.type) {
        case 'structureDefinition/REQUEST':
            return { ...state, isFetching: true, error: undefined, items: defaultState.items };
        case 'structureDefinition/RECEIVE':
            return {
                ...state,
                isFetching: false,
                hasFetched: true,
                error: undefined,
                itemsRequested: action.payload.length
            };
        case 'structureDefinition/SET_ERROR':
            return { ...state, error: 'Error loading structureDefinition.' };
        case 'structureDefinition/REQUEST_IMAGE':
            return { ...state, isFetching: true, error: undefined};
        case 'structureDefinition/RECEIVE_IMAGE':
            const { image, definition } = action.payload;
            // If SVG
            const svg = image.querySelector('svg');
            const box = svg ? svg.viewBox.baseVal : undefined;
            // TODO: What do i do if it's not an svg???????

            const item = {
                ...definition,
                imageWidth: box.width,
                imageHeight: box.height,
                image: svg
            }

            return {
                ...state,
                isFetching: false,
                hasFetched: true,
                error: undefined,
                items: [
                    ...state.items,
                    item
                ],
            };
        default:
            return state;
    }
}


export default structureDefinitionReducer;
