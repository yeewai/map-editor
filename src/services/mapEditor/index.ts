import reducer from "./reducer";
import * as mapEditorActions from "./actions";
import * as mapEditorSelectors from "./selectors";
import * as mapEditorTypes from "./types";

export { mapEditorActions, mapEditorSelectors, mapEditorTypes };
export default reducer;
