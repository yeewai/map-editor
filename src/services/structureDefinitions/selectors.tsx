import { createSelector } from 'reselect';
import _ from 'lodash';

import { StateTree } from 'services/types';

export const get = ( state: StateTree ) => state.structureDefinitions.items;
export const getError = (state: StateTree ) => state.structureDefinitions.error;
export const isCurrentProfileFetching = (state: StateTree ) => state.structureDefinitions.isFetching;
export const getHasFetched = (state: StateTree ) => state.structureDefinitions.hasFetched;

export const getByGroup = createSelector(
    get,
    ( sdefinitions ) => ( _.groupBy(sdefinitions, "kind") )
);
