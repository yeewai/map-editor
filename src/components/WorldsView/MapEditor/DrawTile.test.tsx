import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';

import DrawTile from './DrawTile';
import { structureDefinitionTypes } from 'services/structureDefinitions';

Enzyme.configure({ adapter: new Adapter() });

describe("Drawing the board", () => {
    let sandbox, wrapper, sd;

    beforeEach(() => {
        sandbox = sinon.createSandbox();
        sd = structureDefinitionTypes.emptyStructureDefintion;
    });

    afterEach(() => {
        sandbox.restore();
    });

    it ("renders", () => {
        sd.imageWidth = 1;
        sd.imageHeight = 2;

        wrapper = shallow(<DrawTile structureDefinition={sd} x={1} y={2} />);
        expect(wrapper).toMatchSnapshot();
    });

    it ("renders the null tile if no structure", () => {
        wrapper = shallow(<DrawTile structureDefinition={null} x={1} y={2} />);
        expect(wrapper).toMatchSnapshot();
    })

    it ("doesn't rerender if the structure hasn't changed (eg adjusting the camera)", () => {
        const wrapper = shallow(<DrawTile structureDefinition={sd} />)

        expect(wrapper.instance().shouldComponentUpdate({structureDefinition: sd})).toBe(false)
        expect(wrapper.instance().shouldComponentUpdate({structureDefinition: "WoooOOoOo"})).toBe(true)
    })

});
