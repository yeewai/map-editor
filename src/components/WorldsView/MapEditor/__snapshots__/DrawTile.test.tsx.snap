// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawing the board renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DrawTile
    structureDefinition={
        Object {
            "id": "",
            "imageHeight": 2,
            "imageUrl": "",
            "imageWidth": 1,
            "kind": "",
            "length": 1,
            "name": "",
            "width": 1,
          }
    }
    x={1}
    y={2}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "tile",
      "height": 20,
      "width": 10,
      "x": -10,
      "xlinkHref": "#",
      "y": -5.566243270259357,
    },
    "ref": null,
    "rendered": null,
    "type": "use",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "tile",
        "height": 20,
        "width": 10,
        "x": -10,
        "xlinkHref": "#",
        "y": -5.566243270259357,
      },
      "ref": null,
      "rendered": null,
      "type": "use",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Drawing the board renders the null tile if no structure 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DrawTile
    structureDefinition={null}
    x={1}
    y={2}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "nullTile",
      "points": "-5,8.660254037844386 0,11.547005383792515 -5,14.433756729740644 -10,11.547005383792515",
    },
    "ref": null,
    "rendered": null,
    "type": "polygon",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "nullTile",
        "points": "-5,8.660254037844386 0,11.547005383792515 -5,14.433756729740644 -10,11.547005383792515",
      },
      "ref": null,
      "rendered": null,
      "type": "polygon",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
